packages/workflow/src/WorkflowDataProxy.ts

```diff
@@ -1018,8 +1042,19 @@ export class WorkflowDataProxy {
 										itemIndex = that.itemIndex;
 									}
 
-									const executionData = that.connectionInputData;
-									const input = executionData[itemIndex];
+									const pinnedData = getPinDataIfManualExecution(
+										that.workflow,
+										nodeName,
+										that.mode,
+									);
+									const executionData = that.connectionInputData.length
+										? that.connectionInputData
+										: pinnedData;
+									const input = executionData?.[itemIndex];
+
+									if (pinnedData && !that.connectionInputData.length) {
+										return input;
+									}
```

Comment:
Ideally we should only get pindata when needed (not everytime) but I'm not sure my refactoring attempt is any better.

```ts
const input = that.connectionInputData?.[itemIndex];

if (!input) {
	const pinData = getPinDataIfManualExecution(that.workflow, nodeName, that.mode);
	const pinDataItem = pinData?.[itemIndex];

	if (pinDataItem) return pinDataItem;
}
```